spec/index.md

```diff
@@ -5033,6 +5083,175 @@ S with
 
 ```
 
+#### IC Management Canister: Take snapshot
+
+Only the controllers of the given canister can take a snapshot.
+
+```html
+
+S.messages = Older_messages · CallMessage M · Younger_messages
+(M.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+M.callee = ic_principal
+M.method_name = 'take_snapshot'
+M.arg = candid(A)
+M.caller ∈ S.controllers[A.canister_id]
+
+Cycles_reserved = cycles_to_reserve(S, A.canister_id, S.compute_allocation[A.canister_id], S.memory_allocation[A.canister_id] + memory_usage_snapshot(New_snapshot), S.canisters[A.canister_id])
+New_balance = S.balances[A.canister_id] - Cycles_used - Cycles_reserved
+New_reserved_balance = S.reserved_balances[A.canister_id] + Cycles_reserved
+New_reserved_balance ≤ S.reserved_balance_limits[A.canister_id]
+
+liquid_balance(
+  New_balance,
+  New_reserved_balance,
+  freezing_limit(
+    S.compute_allocation[A.canister_id],
+    S.memory_allocation[A.canister_id],
+    S.freezing_threshold[A.canister_id],
+    memory_usage_wasm_state(New_state) +
+      memory_usage_raw_module(A.wasm_module) +
+      memory_usage_canister_history(New_canister_history) +
+      memory_usage_snapshot(New_snapshot),
+    S.canister_subnet[A.canister_id].subnet_size,
+  )
+) ≥ 0
+
+```
+
+State after  
+
+```html
+
+S with
+    snapshots = Other_snapshots · New_snapshot
+    balances[A.canister_id] = New_balance
+    reserved_balances[A.canister_id] = New_reserved_balance
+    messages = Older_messages · Younger_messages ·
+      ResponseMessage {
+        origin = M.origin;
+        response = Reply (candid());
+        refunded_cycles = M.transferred_cycles;
+      }
+
+```
+
+
+#### IC Management Canister: Load snapshots
+
+
+Only the controllers of the given canister can load a snapshot.
+
+```html
+
+S.messages = Older_messages · CallMessage M · Younger_messages
+(M.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+M.callee = ic_principal
+M.method_name = 'load_snapshot'
+M.arg = candid(A)
+M.caller ∈ S.controllers[S.snapshots[A.snapshot_id].canister_id]
```

Comment:
Above, S.snapshots was indexed by canister ID, here it's indexed by snapshot ID. I suppose, the snapshots will be part of the canister's state in the replica implementation and thus it makes sense to index S.snapshots by canister IDs (and also require a canister ID in the mgmt calls arguments so that the mapping snapshot ID -> canister ID does not need to be computed by the replica separately)

Reply:
I think I forgot top reply to this one. We don't have canister Id here. The snapshots are stored in the replica, not on canister. If you have any idea how to reflect that in the semantics, please let me know. Otherwise, I will consider this as done.